### ISSUE TITLE
0.22 regression: R.uniq is broken for function (reference-eqaulity case)

### ISSUE THREAD

**Issue by asaf-romano** [ramda/ramda#1987]
```js
uniq([identity, identity, identity, identity, identity, identity]).length // 5
```

**Comment by asaf-romano** [comment]
@buzzdecafe

**Comment by CrossEye** [comment]
We have a value-equality implementation.  

If you can find a way to tell that two functions are equal, [Professor Rice][rt] would like to have some words.  :smile:

  [rt]: https://en.wikipedia.org/wiki/Rice's_theorem

**Comment by CrossEye** [comment]
I didn't mean to leave that there.

It's quite reasonable that we extend our equality tests to at least handle reference equality for functions.

**Comment by buzzdecafe** [comment]
curious indeed

```js
uniq([identity, identity]).length //=> 1
uniq([identity, identity, identity]).length //=> 2 WTF
```

**Comment by asaf-romano** [comment]
We always used reference-equality as a fallback there. Anyway, since equals(f1, f1) returns true, you'd expect unique to apply the same equality semantics.

**Comment by buzzdecafe** [comment]
comedy

`uniq([add, identity]) //=> []`

**Comment by CrossEye** [comment]
Cripes.  I'm going to crawl under a rock and hope this goes away.

**Comment by davidchambers** [comment]
> We always used reference-equality as a fallback there.

That approach seems reasonable to me. Here's the definition of [`Function#fantasy-land/equals`][1] from sanctuary-type-classes:

> ```javascript
> //  Function$prototype$equals :: Function ~> Function -> Boolean
> var Function$prototype$equals = function(other) {
>   return other === this;
> };
> ```

I really do think we should add sanctuary-type-classes as a dependency. It would be worthwhile for `equals` and `toString` alone. The fact that we'd also get `concat`, `empty`, `map`, `bimap`, `promap`, `ap`, `chain`, `unnest`, `filter`, `traverse`, and `sequence` is just gravy. :)


[1]: https://github.com/sanctuary-js/sanctuary-type-classes/blob/v1.2.0/index.js#L741-L744

**Comment by CrossEye** [comment]
> I really do think we should add sanctuary-type-classes as a dependency. 

Are you suggesting that this would fix the issue at hand?  I haven't had time to investigate, but was assuming that the issue has to do with [`internal/_Set`][is].  Certainly our current `equals` [handles functions][hf] using reference equality, as do `contains` and `indexOf`.


  [is]: https://github.com/ramda/ramda/blob/v0.22.1/src/internal/_Set.js
  [hf]: https://goo.gl/XSuDoV

**Comment by davidchambers** [comment]
My mistake, @CrossEye. I should have read the comments more carefully before commenting myself.

**Comment by CrossEye** [comment]
@davidchambers: That's not to say I'm objecting to your proposal.  There are plenty of good arguments for that.  I just didn't think this was one.  :wink:

**Comment by scott-christopher** [comment]
Closed via #1992
